collector/logs/sources/tail/tailer.go

```diff
@@ -0,0 +1,201 @@
+}
+
+// Tailer is a specific instance of a file being tailed.
+type Tailer struct {
+	tail           *tail.Tail
+	shutdown       context.CancelFunc
+	errgroup       *errgroup.Group
+	database       string
+	table          string
+	logTypeParser  sourceparse.LogTypeParser
+	logLineParsers []parser.Parser
+	resources      map[string]interface{}
+}
+
+func StartTailing(config TailerConfig) (*Tailer, error) {
+	group, ctx := errgroup.WithContext(context.Background())
+	ctx, shutdown := context.WithCancel(ctx)
+
+	batchQueue := make(chan *types.Log, 512)
+	outputQueue := make(chan *types.LogBatch, 1)
+	tailConfig := tail.Config{Follow: true, ReOpen: true, Poll: true}
+	existingCursorPath := cursorPath(config.CursorDirectory, config.Target.FilePath)
+	fileId, position, err := readCursor(existingCursorPath)
+	if err == nil {
+		if logger.IsDebug() {
+			logger.Debugf("TailSource: found existing cursor for file %q: %s %d", config.Target.FilePath, fileId, position)
+		}
+		tailConfig.Location = &tail.SeekInfo{
+			Offset:         position,
+			Whence:         io.SeekStart,
+			FileIdentifier: fileId,
+		}
+	}
+
+	tailFile, err := tail.TailFile(config.Target.FilePath, tailConfig)
+	if err != nil {
+		shutdown()
+		return nil, fmt.Errorf("addTarget create tailfile: %w", err)
+	}
+
+	parsers, err := parser.NewParsers(config.Target.Parsers)
+	if err != nil {
+		shutdown()
+		return nil, fmt.Errorf("addTarget create parsers: %w", err)
+	}
+
+	attributes := make(map[string]interface{})
+	for k, v := range config.Target.Resources {
+		attributes[k] = v
+	}
+
+	tailer := &Tailer{
+		tail:           tailFile,
+		shutdown:       shutdown,
+		errgroup:       group,
+		database:       config.Target.Database,
+		table:          config.Target.Table,
+		logTypeParser:  sourceparse.GetLogTypeParser(config.Target.LogType),
+		logLineParsers: parsers,
+		resources:      attributes,
+	}
+	group.Go(func() error {
+		return readLines(ctx, tailer, config.UpdateChan, batchQueue)
+	})
+
+	batchConfig := engine.BatchConfig{
+		MaxBatchSize: 1000,
+		MaxBatchWait: 1 * time.Second,
+		InputQueue:   batchQueue,
+		OutputQueue:  outputQueue,
+		AckGenerator: config.AckGenerator,
+	}
+	group.Go(func() error {
+		return engine.BatchLogs(ctx, batchConfig)
+	})
+
+	worker := config.WorkerCreator(config.WorkerName, outputQueue)
+	group.Go(worker.Run)
+
+	return tailer, nil
+}
+
+// Stop stops the tailer and cleans up resources.
+// Does not wait for the tailer to finish processing, to allow closing many tailers concurrently.
+// Call Wait() after calling Stop() to wait for the tailer to finish processing.
+func (t *Tailer) Stop() {
+	t.tail.Cleanup()
+	t.tail.Stop()
+	t.shutdown()
+}
+
+// Wait waits for the tailer to finish processing.
+func (t *Tailer) Wait() {
+	t.errgroup.Wait()
+}
+
+func readLines(ctx context.Context, tailer *Tailer, updateChannel <-chan FileTailTarget, outputQueue chan<- *types.Log) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
```

Comment:
Should we return `ctx.Err()` so that the caller, in this case an `errgroup` correctly terminates all other members of that group?